pkg/pillar/cmd/downloader/mdns.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+// github.com/grandcat/zeroconf: under MIT License
+
+package downloader
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/grandcat/zeroconf"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+// given a local server URL string, and returns the modified url with server IP address in it,
+// and bridge name and source IP address for connecting to the server.
+func findDSmDNS(dctx *downloaderContext, serverURL string) (string, string, net.IP, error) {
+	dsURL, err := url.Parse(serverURL)
+	if err != nil {
+		log.Errorf("findDSmDNS: error %v", err)
+		return "", "", nil, err
+	}
+	hostname := dsURL.Hostname()
+	var newURL string
+	var foundIP []net.IP
+
+	urls := strings.Split(serverURL, hostname)
+	if len(urls) != 2 {
+		return "", "", nil, fmt.Errorf("findDSmDNS: urls format error %v", urls)
+	}
+
+	// try to loop through the well-known services, workstation, http and https
+	// until success, for the local data store zeroconf
+	services := []string{
+		"workstation", "http", "https",
+	}
+
+	sub := dctx.subNetworkInstanceStatus
+	niItems := sub.GetAll()
+	ifs := findLocalBridges(niItems)
+	if len(ifs) == 0 {
+		return "", "", nil, fmt.Errorf("findDSmDNS: has no bridge interface")
+	}
+
+	for _, service := range services {
+		foundIP, err = queryService(ifs, hostname, service)
+		if err != nil {
+			return "", "", nil, err
+		}
+		if len(foundIP) > 0 {
+			break
+		}
+	}
+
+	if len(foundIP) == 0 {
+		return "", "", nil, fmt.Errorf("findDSmDNS: mDNS host not found")
+	}
+
+	// find the first src/dst pair we can use for local DS downloading
```

Comment:
FWIW in some future things might match multiple IP addresses (when we can handle overlapping IP subnets).
I wonder if it would make sense to return a list of tuples.

Reply:
but for this current usage, we only need one. since it's internal connection and we don't have external network problems. this mDNS app can actually return multiple intf ip address here say for both eth0 and eth1 of the App which can both be used to download images from the App. But here when we match the whatever the bridge source to the App address we assigned, we know it will work for downloading. We can revisit this later if new cases comes.

Reply:
OK